netaddr.go

```diff
@@ -163,58 +163,34 @@ func (err parseIPError) Error() string {
 }
 
 // parseIPv4 parses s as an IPv4 address (in form "192.168.0.1").
-func parseIPv4(in string) (ip IP, err error) {
-	s := in
-	var ip4 [4]byte
-
-	for i := 0; i < 4; i++ {
-		var (
-			j   int
-			acc uint16
-		)
-		// Parse one byte of digits. Bail if we overflow, stop at
-		// first non-digit.
-		//
-		// As of Go 1.15, don't try to factor this digit reading into
-		// a helper function. Its complexity is slightly too high for
-		// inlining, which ends up costing +50% in parse time.
-		for j = 0; j < len(s); j++ {
-			if s[j] < '0' || s[j] > '9' {
-				break
-			}
-			acc = (acc * 10) + uint16(s[j]-'0')
-			if acc > 255 {
-				return IP{}, parseIPError{in: in, msg: "IPv4 field has value >255", at: s}
+func parseIPv4(s string) (ip IP, err error) {
+	var ip4 [4]uint8
+	var a, b int
+	for i := 0; i < len(s); i++ {
+		if s[i] >= '0' && s[i] <= '9' {
```

Comment:
Playing with it I have found that subtracting `'0'` first and converting to uint eliminates one `cmp` from assembly and it only checks if value is below 9.

Reply:
The compiler should do this for you if you assign s[i] to a variable first (so it can see that it can safely re-evaluate the expression multiple times). That may be a bit more readable.

Reply:
Another way to do this is to do something like:

```go
for i, b := range []byte(s) {
```

Reply:
And for a fun twist, I believe the compiler will generate the same code if you do:

```go
switch s[i] {
case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
  // etc
case '.':
  // etc
default:
  // etc
}
```